include/xparrow/xbuffer.hpp

```diff
@@ -0,0 +1,239 @@
+/***************************************************************************
+* Copyright (c) QuantStack                                                 *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+***************************************************************************/
+
+#pragma once
+
+#include <algorithm>
+#include <concepts>
+#include <cstdint>
+
+namespace xp
+{
+    namespace impl
+    {
+        template <class T>
+        struct xbuffer_data
+        {
+            using value_type = T;
+            using pointer = T*;
+            using size_type = std::size_t;
+            
+            bool empty() const noexcept;
+            size_type size() const noexcept;
+
+            template <class U = T>
+            U* data() noexcept;
+
+            template <class U = T>
+            const U* data() const noexcept;
+
+            void swap(xbuffer_data& rhs) noexcept;
+            bool equal(const xbuffer_data& rhs) const;
+
+            pointer p_data = nullptr;
+            size_type m_size = 0;
+        };
+    }
+
+    /**
+     * @class xbuffer
+     * @brief Object that owns a piece of contiguous memory
+     */
+    template <class T>
+    class xbuffer : private impl::xbuffer_data<T>
+    {
+    public:
+
+        using base_type = impl::xbuffer_data<T>;
+        using value_type = typename base_type::value_type;
+        using pointer = typename base_type::pointer;
+        using size_type = typename base_type::size_type;
+
+        xbuffer() = default;
+        explicit xbuffer(size_type size);
+        xbuffer(pointer data, size_type size);
+        
+        ~xbuffer();
+        
+        xbuffer(const xbuffer&);
+        xbuffer& operator=(const xbuffer&);
+
+        xbuffer(xbuffer&&);
+        xbuffer& operator=(xbuffer&&);
+
+        using base_type::empty;
+        using base_type::size;
+        using base_type::data;
+
+        void resize(size_type new_size);
+
+        void swap(xbuffer&) noexcept;
+        bool equal(const xbuffer& rhs) const;
+
+    private:
+
+        pointer allocate(size_type size) const;
+        void deallocate(pointer mem) const;
+    };
+
+    template <class T>
+    bool operator==(const xbuffer<T>& lhs, const xbuffer<T>& rhs);
+
+    /**
+     * @class xbuffer_reference
+     * @brief Object that wraps a piece of contiguous memory
+     * but does not own it.
+     */
+    /*class xbuffer_reference : private xbuffer_data
```

Comment:
Why is this commented? Or should it be removed?

Reply:
It should be added later actually. That' a leftover that I can remove.